### PR TITLE
WIP: src/output: add wasapi output and mixer plugin

### PR DESCRIPTION
**Pull request by ibmibmibm**
Add WASAPI event output plugin with doc.

### CODE REVIEW

**Comment by ibmibmibm** [comment]
Waiting for #883

**Comment by MaxKellermann** [comment]
#883 is waiting for #883? Hm?

File: src/win32/Com.hxx [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#ifndef MPD_WIN32_COM_HXX
+#define MPD_WIN32_COM_HXX
+
+#include <objbase.h>
+#include <windows.h>
+
+class COM {
```
Comment (MaxKellermann) [inline src/win32/Com.hxx:26]: Badly named and undocument - what does this class do? COM what?

File: src/win32/Com.hxx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+public:
+	COM() noexcept : result(CoInitializeEx(nullptr, COINIT_MULTITHREADED)) {}
```
Comment (MaxKellermann) [inline src/win32/Com.hxx:28]: I'd prefer to throw an exception if `CoInitializeEx()` fails. I don't like partially-initialized objects, and exceptions are the right tool for such a job.

File: src/win32/Com.hxx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ */
+
+#ifndef MPD_WIN32_COM_HXX
+#define MPD_WIN32_COM_HXX
+
+#include <objbase.h>
+#include <windows.h>
+
+class COM {
+public:
```
Comment (MaxKellermann) [inline src/win32/Com.hxx:27]: This class is copyable - better prevent this.

File: src/win32/Com.hxx [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	~COM() noexcept {
+		if (*this) {
+			CoUninitialize();
+		}
+	}
+	operator bool() noexcept { return SUCCEEDED(result); }
+	HRESULT Result() const { return result; }
+
+protected:
+	HRESULT result = 0;
```
Comment (MaxKellermann) [inline src/win32/Com.hxx:38]: This initializer is pointless and harmful. Don't initialize variables with arbitrary values - this only hides initialization errors from valgrind.

File: src/win32/ComHeapPtr.hxx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+#ifndef MPD_WIN32_COMHEAPPTR_HXX
+#define MPD_WIN32_COMHEAPPTR_HXX
+
+#include <objbase.h>
+#include <utility>
+#include <windows.h>
+
+template <typename T>
+class ComHeapPtr {
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:28]: API documentation, please! What does this class do?

File: src/win32/ComHeapPtr.hxx [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+public:
+	ComHeapPtr() noexcept = default;
+	ComHeapPtr(std::nullptr_t) noexcept {}
+	ComHeapPtr(const ComHeapPtr &rhs) noexcept = delete;
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:32]: If a move constructor exists, then this line is redundant. You can state that explicitly and keep it if you prefer to - I usually prefer to not to, so it boils down to personal taste.

File: src/win32/ComHeapPtr.hxx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	ComHeapPtr(ComHeapPtr &&rhs) noexcept : ptr(rhs.ptr) { rhs.ptr = nullptr; }
+	ComHeapPtr &operator=(ComHeapPtr &rhs) noexcept = delete;
+	ComHeapPtr &operator=(ComHeapPtr &&rhs) noexcept {
+		std::swap(ComHeapPtr(std::move(rhs)), *this);
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:36]: This looks like a rather convoluted way of saying `std::swap(ptr, rhs.ptr)`, doesn't it?

File: src/win32/ComHeapPtr.hxx [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+			CoTaskMemFree(p);
+		}
+	}
+	constexpr bool operator!() const noexcept { return !ptr; }
+	constexpr T **operator&() noexcept { return &ptr; }
+	constexpr T &operator*() const noexcept { return *ptr; }
+	constexpr T *operator->() const noexcept { return ptr; }
+	constexpr bool operator==(T *rhs) const noexcept { return ptr == rhs; }
+	constexpr bool operator!=(T *rhs) const noexcept { return ptr != rhs; }
+	constexpr bool operator<(T *rhs) const noexcept { return ptr < rhs; }
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:58]: What's the point of this operator? Why would one ever compare pointer values this way?

File: src/win32/ComHeapPtr.hxx [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	constexpr T *Detach() noexcept { return std::exchange(ptr, nullptr); }
+	void Free() noexcept {
+		T *p = nullptr;
+		std::swap(ptr, p);
+		if (p) {
+			CoTaskMemFree(p);
+		}
+	}
+	constexpr bool operator!() const noexcept { return !ptr; }
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:52]: Redundant operator - you get this for free by implementing `operator bool()`.

File: src/win32/ComHeapPtr.hxx [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+	constexpr T **operator&() noexcept { return &ptr; }
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:53]: What's the point of this operator?

File: src/win32/ComHeapPtr.hxx [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+	}
+	~ComHeapPtr() noexcept { Free(); }
+	void Attach(T *new_ptr) noexcept {
+		Free();
+		ptr = new_ptr;
+	}
+	constexpr T *Detach() noexcept { return std::exchange(ptr, nullptr); }
+	void Free() noexcept {
+		T *p = nullptr;
+		std::swap(ptr, p);
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:47]: Convoluted way of saying `std::exchange(ptr, nullptr)`.

File: src/win32/ComHeapPtr.hxx [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+	ComHeapPtr(std::nullptr_t) noexcept {}
+	ComHeapPtr(const ComHeapPtr &rhs) noexcept = delete;
+	ComHeapPtr(ComHeapPtr &&rhs) noexcept : ptr(rhs.ptr) { rhs.ptr = nullptr; }
+	ComHeapPtr &operator=(ComHeapPtr &rhs) noexcept = delete;
+	ComHeapPtr &operator=(ComHeapPtr &&rhs) noexcept {
+		std::swap(ComHeapPtr(std::move(rhs)), *this);
+		return *this;
+	}
+	~ComHeapPtr() noexcept { Free(); }
+	void Attach(T *new_ptr) noexcept {
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:40]: Can't we follow STL conventions here and call it `reset()`?

File: src/win32/ComHeapPtr.hxx [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+		Free();
+		ptr = new_ptr;
+	}
+	constexpr T *Detach() noexcept { return std::exchange(ptr, nullptr); }
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:44]: STL: `release()`

File: src/win32/ComHeapPtr.hxx [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+		if (p) {
+			CoTaskMemFree(p);
+		}
+	}
+	constexpr bool operator!() const noexcept { return !ptr; }
+	constexpr T **operator&() noexcept { return &ptr; }
+	constexpr T &operator*() const noexcept { return *ptr; }
+	constexpr T *operator->() const noexcept { return ptr; }
+	constexpr bool operator==(T *rhs) const noexcept { return ptr == rhs; }
+	constexpr bool operator!=(T *rhs) const noexcept { return ptr != rhs; }
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:57]: What's the point of these operators?

File: src/win32/ComHeapPtr.hxx [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+	constexpr bool operator<(T *rhs) const noexcept { return ptr < rhs; }
+	constexpr operator const T *() const noexcept { return ptr; }
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:59]: This operator is inconsistent. The `operator*` and `operator->` are both `const` and will happily return a non-`const` pointer. So what's your design policy - does a `const` instance only give `const` pointers or not?

File: src/win32/ComHeapPtr.hxx [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+	constexpr operator T *() noexcept { return ptr; }
+
+	template <typename U = void>
+	U *Cast() noexcept {
+		return reinterpret_cast<U *>(ptr);
+	}
+
+	template <typename U = void>
+	U **AddressCast() noexcept {
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:68]: Why?

File: src/win32/ComHeapPtr.hxx [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+		return reinterpret_cast<U **>(&ptr);
+	}
+
+	void swap(ComHeapPtr &rhs) noexcept { std::swap(ptr, rhs.ptr); }
```
Comment (MaxKellermann) [inline src/win32/ComHeapPtr.hxx:72]: This should be a `friend` and not a method, shouldn't it?

File: src/win32/ComPtr.hxx [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+	ComPtr(const ComPtr &rhs) noexcept : ptr(rhs.ptr) {
+		if (ptr) {
+			ptr->AddRef();
+		}
+	}
+	ComPtr(ComPtr &&rhs) noexcept : ptr(rhs.ptr) { rhs.ptr = nullptr; }
+	~ComPtr() noexcept { Release(); }
+
+	ComPtr &operator=(ComPtr<T> &rhs) noexcept {
+		std::swap(ComPtr<T>(rhs), *this);
```
Comment (MaxKellermann) [inline src/win32/ComPtr.hxx:41]: This is an obscure way of creating a new refernce.

File: src/win32/ComPtr.hxx [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+
+#ifndef MPD_WIN32_COMPTR_HXX
+#define MPD_WIN32_COMPTR_HXX
+
+#include <objbase.h>
+#include <utility>
+#include <windows.h>
+
+template <typename T>
+class ComPtr {
```
Comment (MaxKellermann) [inline src/win32/ComPtr.hxx:28]: Again, documentation missing. What is this?

File: src/win32/ComPtr.hxx [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+	}
+	ComPtr(ComPtr &&rhs) noexcept : ptr(rhs.ptr) { rhs.ptr = nullptr; }
+	~ComPtr() noexcept { Release(); }
+
+	ComPtr &operator=(ComPtr<T> &rhs) noexcept {
+		std::swap(ComPtr<T>(rhs), *this);
+		return *this;
+	}
+	ComPtr &operator=(ComPtr<T> &&rhs) noexcept {
+		std::swap(ComPtr<T>(std::move(rhs)), *this);
```
Comment (MaxKellermann) [inline src/win32/ComPtr.hxx:45]: More obscure code. Why not `std::swap(ptr, rhs.ptr)`?

File: src/win32/HResult.hxx [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+#include <cstdarg>
+#include <cstdio>
+#include <stdexcept>
+#include <string_view>
+#include <system_error>
+#include <vector>
+
+#include <audiopolicy.h>
+
+gcc_const constexpr std::string_view HRESULTToString(HRESULT result) {
```
Comment (MaxKellermann) [inline src/win32/HResult.hxx:35]: `gcc_const` is redundant when you already have `constexpr` (the latter is a superset).

File: src/win32/WinEvent.hxx [hunk 22]
```diff
@@ -0,0 +1,50 @@
...
+
+class WinEvent {
+public:
+	WinEvent() : event(CreateEventW(nullptr, false, false, nullptr)) {
+		if (!event) {
+			throw FormatLastError("Error creating events");
+		}
+	}
+	~WinEvent() noexcept { CloseHandle(event); }
+	WinEvent(WinEvent &&) = default;
```
Comment (MaxKellermann) [inline src/win32/WinEvent.hxx:34]: Does this really work?

File: src/win32/WinEvent.hxx [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+	WinEvent(const WinEvent &) = delete;
+	WinEvent &operator=(WinEvent &&) = default;
```
Comment (MaxKellermann) [inline src/win32/WinEvent.hxx:36]: And this?

**Comment by ibmibmibm** [comment]
These interface in ComPtr and ComHeapPtr is a mimic to Microsoft ATL.
https://docs.microsoft.com/zh-tw/cpp/atl/active-template-library-atl-concepts?view=vs-2019

**Comment by MaxKellermann** [comment]
> I can remove those functions that not used, or change interfaces to more STL like if you ask.

If this follows ATL coding style (and Windows developers are usually familiar with ATL), then this sounds like a reasonable choice. (Assuming that ATL is a well-designed library, which I don't know.)

I leave the decision to you, hoping you'll keep maintaining your code - but code that doesn't make much sense (like some of those operators I criticized) should better be removed unless it is really used or useful for something.

**Comment by ibmibmibm** [comment]
> 
> 
> > I can remove those functions that not used, or change interfaces to more STL like if you ask.
> 
> If this follows ATL coding style (and Windows developers are usually familiar with ATL), then this sounds like a reasonable choice. (Assuming that ATL is a well-designed library, which I don't know.)
> 
> I leave the decision to you, hoping you'll keep maintaining your code - but code that doesn't make much sense (like some of those operators I criticized) should better be removed unless it is really used or useful for something.

Of course I like STL's design more, so I'll change the interface to STL style LOL

File: src/win32/WinEvent.hxx (hunk 23 above)
Comment (ibmibmibm) [reply src/win32/WinEvent.hxx]: copy/move constructor and copy/move assignment is now deleted.

File: src/win32/WinEvent.hxx (hunk 22 above)
Comment (ibmibmibm) [reply src/win32/WinEvent.hxx]: copy/move constructor and copy/move assignment is now deleted.

File: src/win32/ComPtr.hxx (hunk 20 above)
Comment (ibmibmibm) [reply src/win32/ComPtr.hxx]: Done.

File: src/win32/HResult.hxx (hunk 21 above)
Comment (ibmibmibm) [reply src/win32/HResult.hxx]: Done.

File: src/win32/ComPtr.hxx (hunk 19 above)
Comment (ibmibmibm) [reply src/win32/ComPtr.hxx:32]: Brief documents added.

File: src/win32/ComPtr.hxx (hunk 18 above)
Comment (ibmibmibm) [reply src/win32/ComPtr.hxx]: Now use AddRef.

File: src/win32/ComHeapPtr.hxx (hunk 15 above)
Comment (ibmibmibm) [reply src/win32/ComHeapPtr.hxx]: Removed.

File: src/win32/ComHeapPtr.hxx (hunk 14 above)
Comment (ibmibmibm) [reply src/win32/ComHeapPtr.hxx]: Removed.

File: src/win32/ComHeapPtr.hxx (hunk 13 above)
Comment (ibmibmibm) [reply src/win32/ComHeapPtr.hxx]: Changed.

File: src/win32/ComHeapPtr.hxx (hunk 12 above)
Comment (ibmibmibm) [reply src/win32/ComHeapPtr.hxx]: Changed.

File: src/win32/ComHeapPtr.hxx (hunk 11 above)
Comment (ibmibmibm) [reply src/win32/ComHeapPtr.hxx]: Fixed.

File: src/win32/ComHeapPtr.hxx (hunk 10 above)
Comment (ibmibmibm) [reply src/win32/ComHeapPtr.hxx]: Replaced with `Address()`.

File: src/win32/ComHeapPtr.hxx (hunk 9 above)
Comment (ibmibmibm) [reply src/win32/ComHeapPtr.hxx]: Removed.

File: src/win32/ComHeapPtr.hxx (hunk 8 above)
Comment (ibmibmibm) [reply src/win32/ComHeapPtr.hxx]: Removed.

File: src/win32/ComHeapPtr.hxx (hunk 7 above)
Comment (ibmibmibm) [reply src/win32/ComHeapPtr.hxx]: Changed.

File: src/win32/ComHeapPtr.hxx (hunk 5 above)
Comment (ibmibmibm) [reply src/win32/ComHeapPtr.hxx:32]: Brief documents added.

File: src/win32/Com.hxx (hunk 4 above)
Comment (ibmibmibm) [reply src/win32/Com.hxx]: Removed.

File: src/win32/Com.hxx (hunk 3 above)
Comment (ibmibmibm) [reply src/win32/Com.hxx:30]: copy/move constructor and copy/move assignment disabled.

File: src/win32/Com.hxx (hunk 2 above)
Comment (ibmibmibm) [reply src/win32/Com.hxx]: Error throwing code are moved here now.

File: src/win32/Com.hxx (hunk 1 above)
Comment (ibmibmibm) [reply src/win32/Com.hxx:29]: Brief documents added.

**Comment by MaxKellermann** [comment]
Thanks, I've merged this, even though I'm not 100% happy with some implementation details - for example the method `WasapiOutput::EnumerateDevices()` which instead of returning a list of devices, populates a `std::vector` field. Some things could be made more robust and/or easier, but in any case it's good to have your code, and we can improve it any time.
What I'd like to see is an explanation for users (and for Windows ignorants like me) why there are now two output plugins for Windows; what's the difference, and under which circumstances would you choose one or the other? That could/should be part of the MPD documentation (the protocol reference?).
Since you appear to be interested in MPD's Windows support - the MPD documentation is very scarce on Windows installations instructions; maybe that deserves a dedicates section in the manual. Most users don't understand where to put `mpd.conf`, and maybe MPD could be integrated better into Windows. Any work welcome!

**Pull request merged**